src/relations/async_replication.py

```diff
@@ -0,0 +1,291 @@
+
+        # Check if we have already published pod-address. If not, then we are waiting
+        # for the leader to catch all the pod ips and restart itself
+        if "pod-address" not in replica_relation.data[self.charm.unit]:
+            replica_relation.data[self.charm.unit]["pod-address"] = _get_pod_ip()
+            # Finish here and wait for the retrigger from the primary cluster
+            return
+
+        self.container.stop(self.charm._postgresql_service)
+        # Standby units must delete their data folder
+        # Delete the K8S endpoints that tracks the cluster information, including its id.
+        # This is the same as "patronictl remove patroni-postgresql-k8s", but the latter doesn't
+        # work after the database service is stopped on Pebble.
+        try:
+            client = Client()
+            client.delete(
+                Endpoints,
+                name=f"patroni-{self.charm._name}",
+                namespace=self.charm._namespace,
+            )
+            client.delete(
+                Endpoints,
+                name=f"patroni-{self.charm._name}-config",
+                namespace=self.charm._namespace,
+            )
+
+            self.container.exec("rm -r /var/lib/postgresql/data/pgdata".split()).wait_output()
+            self.charm._create_pgdata(self.container)
+
+            self.charm.update_config()
+        except Exception:
+            pass
+        self.container.start(self.charm._postgresql_service)
+
+    def _get_primary_candidates(self):
+        rel = self.model.get_relation(ASYNC_PRIMARY_RELATION)
+        return rel.units if rel else []
+
+    def _check_if_primary_already_selected(self) -> Unit:
+        """Returns the unit if a primary is present."""
+        result = None
+        if not self.relation_set:
+            return None
+        for rel in self.relation_set:
+            for unit in self._all_units(rel):
+                if "elected" in rel.data[unit] and not result:
+                    result = unit
+                elif "elected" in rel.data[unit] and result:
+                    raise MoreThanOnePrimarySelectedError
+        return result
+
+    def _on_promote_standby_cluster(self, event: ActionEvent) -> None:
+        """Moves a standby cluster to a primary, if none is present."""
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.fail("Cluster not initialized yet.")
+            return
+
+        if not self.charm.unit.is_leader():
+            event.fail("Not the charm leader unit.")
+            return
+
+        # Let the exception error the unit
+        unit = self._check_if_primary_already_selected()
+        if unit:
+            event.fail(f"Cannot promote - {unit.name} is already primary: demote it first")
+            return
+
+        # If this is a standby-leader, then execute switchover logic
+        # TODO
```

Comment:
Looking at [Patroni docs](https://patroni.readthedocs.io/en/latest/ha_multi_dc.html#asynchronous-replication):
```
Automatic promotion is not possible, because DC2 will never able to figure out the state of DC1.

You should not use pg_ctl promote in this scenario, you need “manually promote” the healthy cluster by removing standby_cluster section from the [dynamic configuration](https://patroni.readthedocs.io/en/latest/dynamic_configuration.html#dynamic-configuration).
```
It sounds to me that we need to remove the new relation (with `--force`?) to promote the replica cluster and we can't really promote and demote with actions.

Reply:
Hi @dragomirp, the way I see the fail/switchover happening is the following:
```
# First, offer async-primary from both models:
juju switch psql-1
juju offer postgresql-k8s:async-primary async-primary

juju switch psql-2
juju offer postgresql-k8s:async-primary async-primary
```

Then, we consume in each model, the async-primary relation:

```
juju consume admin/psql-2.async-primary 
juju relate -m psql-1 postgresql-k8s:async-replica async-primary

juju consume admin/psql-1.async-primary 
juju relate -m psql-2 postgresql-k8s:async-replica async-primary
```

Once that setup is done, the postgresql apps *know* there is an async replication available, but will not implement the actual configuration. That will happen once we run the `promote-standby-cluster` action on one of the models.
At that moment, then the model where the action ran should take over and become the primary. The remaining will continue as replicas.

At switchover, the user must initiate the process. That should be:
```
juju run -m <model-with-old-primary> postgresql-k8s/leader demote-primary-cluster
juju run -m <mode-with-new-primary> postgresql-k8s/leader promote-standby-cluster
```

The demote should not be successful if the target unit sees a cluster as "primary" still connected in its async-replica relation.

In case of failover, I think you are correct, depending on the state of the primary cluster, we may end up with Juju knowing about the primary cluster and having its databag (with the primary key set); but the cluster is gone. Indeed, we would need here to pull the relation out first, possibly with `--force`, then promote one of the replica clusters as leader.

Reply:
But in any case, I don't think we should do automatic failover between clusters, as this is async replication. It should be a conscious decision from the user.

Reply:
It seems that it's important to break the relation or use another mean to disconnect one cluster from the other before promoting, as stated in another part of the docs Dragomir mentioned:

> Before promoting standby cluster one have to manually ensure that the source cluster is down (STONITH). When DC1 recovers, the cluster has to be converted to a standby cluster.